packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/findChildSpanIds.tsx

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) 2023 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { Span } from '../../../../types/trace';
+
+// This function finds child spans for each span and also sorts childSpanIds based on endTime
+const findChildSpanIds = (spans: Span[]): Span[] => {
+  const refinedSpanData: Span[] = [];
+  spans.forEach(span => {
+    if (span.hasChildren) {
```

Comment:
Are we doing redundant work here? if there span already has a `hasChildren` field then I assume the same search logic as already been executed somewhere.

Reply:
Yes you are correct but here there is a small doubt. In [trace.js](https://github.com/jaegertracing/jaeger-ui/blob/main/packages/jaeger-ui/src/selectors/trace.js) file we are counting both child_of and follow_form as a child and 2nd thing is they are sorted based on start time (But we need to sort them based on endTIme for finding Last finishing child of a span). 
So my opinion is even if we remove findChildSpanIds function but we still want to introduce a new function to sort the child spans based on endTime(as in node.children there is only child span Ids not all details of child spans).
What you will say about this??
![a](https://github.com/jaegertracing/jaeger-ui/assets/116095646/47b4feec-89f5-452c-aa84-b965a1916ea1)
![b](https://github.com/jaegertracing/jaeger-ui/assets/116095646/ec7294cf-541b-40c6-bbad-585bfcac9527)

Reply:
Tip: use GitHub permalinks to point / embed source code, not the screenshots. E.g. https://github.com/jaegertracing/jaeger-ui/blob/af8cf8469629f116cf8b1527ca29979b366ec23e/packages/jaeger-ui/src/selectors/trace.js#L33-L64

Reply:
Is seems you can easily utilize that code from `trace.js`:
  * it does the sorting, but the sorting is not inherently the part of building the tree. I would refactor that function to take the comparator as an optional argument (defaulted to the comparator it already uses internally), this way you can ask it to build a tree with children sorted by different criteria
  * another important aspect is that `getTraceSpanIdsAsTree` does not mess up with the trace itself. You're adding children to the main data model, even though you have very specific needs and interpretation of "children" for your algorithm. In contrast, `getTraceSpanIdsAsTree` just builds a separate data structure that you could use in conjunction with the trace itself, which is a better way.

Reply:
If I use sorts the children based on endTime then unfortunately unit tests are failing as already mentoined sorting based on start time is necessary 
https://github.com/jaegertracing/jaeger-ui/blob/main/packages/jaeger-ui/src/model/transform-trace-data.tsx#L119-L120

Reply:
my first bullet point explains how you can have both sorting orders